### ISSUE TITLE
Refactor some connection API calls into flags

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#1190]
The current approach of having to explicitly call `BeginOrContinueRemoteTransaction`, `RemoteTransactionBeginIfNecessary`, and/or `MarkRemoteTransactionCritical` on connections is a little confusing and error-prone.

A more understandable approach would be to support a flag in `GetPlacementConnection` for obtaining a connection with an already-open transaction block, e.g. `GetPlacementConnection(FOR_DML | IN_TRANSACTION, nodename, nodeport)`, which automatically ensures that the connection has a transaction block open that gets committed at the end. We should also automatically call `BeginOrContinueRemoteTransaction` if the `IN_TRANSACTION` flag is set.

The flag would only be supported by `GetPlacementConnection`, not by `StartPlacementConnection`. The latter should probably error out if it is set to avoid cases in which we forget to send BEGIN. Alternatively, we could automatically send BEGIN; by prepending it to the first command that is sent over an IN_TRANSACTION connection, which would allow us to also support `StartPlacementConnection` and would save a round-trip, though this is more complex to implement.

Similarly, we could avoid the need for calling `MarkTransactionCritical` by having a flag like `GetPlacementConnection(CRITICAL | FOR_DML | IN_TRANSACTION, nodename, nodeport)`. This would also be supported by StartPlacementConnection.

**Comment by lithp** [comment]
Some more discussion is here: https://github.com/citusdata/citus/issues/1188

**Comment by anarazel** [comment]
I can live with the change for CRITICAL, but I quite strongly think its entirely the wrong direction for IN_TRANSACTION. We should be moving *more* towards Start*Connection, and having a separate way to start transaction in that case isn't making things any easier to get right.

> The flag would only be supported by GetPlacementConnection, not by StartPlacementConnection. The latter should probably error out if it is set to avoid cases in which we forget to send BEGIN. Alternatively, we could automatically send BEGIN; by prepending it to the first command that is sent over an IN_TRANSACTION connection, which would allow us to also support StartPlacementConnection and would save a round-trip, though this is more complex to implement.

That'd be hard to implement for various reasons. For one, a bunch of callers check the connection state, for anther, there's various cases where you can't just preprend to an existing statements. E.g prepared / parametrized statements, or when row-by-row mode is in effect.

**Comment by marcocitus** [comment]
> That'd be hard to implement for various reasons. For one, a bunch of callers check the connection state, for anther, there's various cases where you can't just preprend to an existing statements.

Not really, you can just keep a "queue" which gets flushed either as a separate execution step or by concatenating commands if possible. This would generally be a good thing to have and has significantly more performance benefit than using Start*Connection.
